kotest-framework/kotest-framework-engine/src/jvmMain/kotlin/io/kotest/engine/spec/runners/SingleInstanceSpecRunner.kt

```diff
@@ -45,12 +47,14 @@ internal class SingleInstanceSpecRunner(
       logger.log { Pair(spec::class.bestName(), "executing spec $spec") }
       try {
          return coroutineScope {
-            pipeline.execute(spec) {
-               val rootTests = materializer.materialize(spec)
-               logger.log { Pair(spec::class.bestName(), "Launching ${rootTests.size} root tests on $scheduler") }
-               scheduler.schedule({ runTest(it, coroutineContext, null) }, rootTests)
-               Result.success(results)
-            }
+            async { // fresh coroutine for each spec
+               pipeline.execute(spec) {
+                  val rootTests = materializer.materialize(spec)
+                  logger.log { Pair(spec::class.bestName(), "Launching ${rootTests.size} root tests on $scheduler") }
+                  scheduler.schedule({ runTest(it, null) }, rootTests)
+                  Result.success(results)
+               }
+            }.await()
```

Comment:
```suggestion
            pipeline.execute(spec) {
               val rootTests = materializer.materialize(spec)
               logger.log { Pair(spec::class.bestName(), "Launching ${rootTests.size} root tests on $scheduler") }
               scheduler.schedule({ runTest(it, null) }, rootTests)
               Result.success(results)
            }
```

The enclosing `coroutineScope` already does create a new coroutine. We don't gain anything when we create another one inside, and `async { ... }.await()` is effectively identical to `coroutineScope { ... }`.

Reply:
Are you sure coroutineScope makes a new coroutine?
Because if I do this:

```
suspend fun main() {
   System.setProperty("kotlinx.coroutines.debug", "on")
   println("Current coroutine $coroutineContext")
   coroutineScope {
      println("Current coroutine $coroutineContext")
      launch {
         println("Current coroutine $coroutineContext")
         coroutineScope {
            println("Current coroutine $coroutineContext")
            launch {
               println("Current coroutine $coroutineContext")
            }
         }
      }
   }
}
```

Only the launch causes the coroutine id to be increased. Maybe coroutineScope makes a new one but keeps the same id ?

Reply:
Yes `coroutineScope` does create a coroutine, namely a `ScopeCoroutine`.

We should not rely on `CoroutineId` providing any meaningful value. It is internal and not documented properly. There are lots of coroutine types, and only some of them (like from `launch`, `async`) do get a new `CoroutineId`.

The easiest way to differentiate between coroutines is via their `Job` context element. If you change the above code to print out just `${coroutineContext[Job]}`, each println will show a new object. And the job hierarchy is what's relevant in terms of waiting and cancellation.

For all practical purposes, I'd say we can safely think of `Job` and coroutine (`AbstractCoroutine` derivative) being equivalent. I'm not sure whether the complicated differentiation between those two inside `kotlinx.coroutines` is still necessary or a historical relic (you know, those in the museum basement), but we should probably not care too much.

On a side note: Use `async { ... }` only if you're interested in the block's result. Otherwise `launch { ... }` is the idiomatic abstraction, along with `join()` to wait for it.

Reply:
Ok that makes sense. So I guess withContext makes a new coroutine too, unless the context isn't actually changing?